source/wifi/wifi_hal.c

```diff
@@ -6700,10 +6722,175 @@ INT wifi_setRadioStatsEnable(INT radioIndex, BOOL enable)
+
+   nl.id = initNl80211(&nl);
+
+
+   if (nl.id < 0) {
+       fprintf(stderr, "Error initializing netlink \n");
+       return -1;
+   }
+
+   struct nl_msg* msg = nlmsg_alloc();
+
+   if (!msg) {
+       fprintf(stderr, "Failed to allocate netlink message.\n");
+       nlfree(&nl);
+       return -2;
+   }
+
+   genlmsg_put(msg,
+              NL_AUTO_PORT,
+              NL_AUTO_SEQ,
+              nl.id,
+              0,
+              0,
+              NL80211_CMD_GET_STATION,
+              0);
+
+   if(mac_addr_aton(mac_addr, clientMacAddress)) {
+       printf("invalid mac address\n");
+       return 0;
+   }
+
+   nla_put(msg, NL80211_ATTR_MAC, MAC_ALEN, mac_addr);
+   nla_put_u32(msg, NL80211_ATTR_IFINDEX, if_nametoindex(if_name));
+   nl_cb_set(nl.cb,NL_CB_VALID,NL_CB_CUSTOM,tidStats_callback,tid_stats);
+   nl_send_auto(nl.socket, msg);
+   nl_recvmsgs(nl.socket, nl.cb);
+   nlmsg_free(msg);
+   nl_cb_put(nl.cb);
+   nlfree(&nl);
+   return RETURN_OK;
+#else
+//iw implementation
+#define TID_STATS_FILE "/tmp/tid_stats_file.txt"
+#define TOTAL_MAX_LINES 50
+
+    char buf[256] = {'\0'}; /* or other suitable maximum line size */
+    char if_name[10]= "wlan0";
+    FILE *fp=NULL;
+    char pipeCmd[1024]= {'\0'};
+    int lines,tid_index=0;
+    char mac_addr[20] = {'\0'};
+
+    wifi_associated_dev_tid_entry_t *stats_entry;
+
+    snprintf(if_name,sizeof(if_name),"wlan%d",radioIndex);
+
+    strcpy(mac_addr,clientMacAddress);
+
+    snprintf(pipeCmd,sizeof(pipeCmd),"iw dev %s station dump -v > "TID_STATS_FILE,if_name);
+    fp= popen(pipeCmd,"r");
+
+    if(fp == NULL)
+        return RETURN_ERR;
+
+    snprintf(pipeCmd,sizeof(pipeCmd),"grep -n 'Station' "TID_STATS_FILE " | cut -d ':' -f1  | head -2 | tail -1");
+    fp=popen(pipeCmd,"r");
+
+    if(fp == NULL)
+        return RETURN_ERR;
+    else if(fgets(buf,sizeof(buf),fp) != NULL)
+        lines=atoi(buf);
+    else
+        return RETURN_ERR; // No devices are connected
+
+    if(lines == 1)
+        lines = TOTAL_MAX_LINES; //only one client is connected , considering next MAX lines of iw output
+
+
+    for(tid_index=0; tid_index<=16; tid_index++)
+    {
+        stats_entry = &tid_stats->tid_array[tid_index];
+        stats_entry->tid = tid_index;
+
+        snprintf(pipeCmd, sizeof(pipeCmd),"cat "TID_STATS_FILE" | awk '/%s/ {for(i=0; i<=%d; i++) {getline; print}}'  |  grep -F -A%d 'MSDU'  | awk '{print $3}' | tail -1",mac_addr,lines,tid_index+2);
+
+        fp=popen(pipeCmd,"r");
+
+        if(fp ==NULL)
+            return RETURN_ERR;
+
+        else if(fgets(buf,sizeof(buf),fp) != NULL)
+            stats_entry->num_msdus = atol(buf);
+
+        stats_entry->ac = _tid_ac_index_get[tid_index];
+//      TODO:
+//      ULLONG ewma_time_ms;    <! Moving average value based on last couple of transmitted msdus
+//      ULLONG sum_time_ms; <! Delta of cumulative msdus times over interval
+    }
+    return RETURN_OK;
+#endif
 }
```

Comment:
I would like to have test caller from main function. Do we have provision for any one test this particular API?

Reply:
I can give you my test code, 

Place the below content into main

int radioIndex=0;
u64 handle= 0;
char sta2[] = "08:7f:98:97:c8:17";
ac_address_t *st2;
st2=sta2;

wifi_associated_dev_tid_stats_t tid_stats;
wifi_getApAssociatedDeviceTidStatsResult(radioIndex,st2,&tid_stats,handle);